bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java

```diff
@@ -12,156 +12,405 @@
-                        HomekitAccessoryType.CARBON_MONOXIDE_SENSOR, itemRegistry)
-                                .orElseThrow(() -> new HomekitException(
-                                        "Carbon monoxide accessory group should have a carbon monoxide sensor in it"));
-                return new HomekitSmokeSensorImpl(carbonMonoxideSensorAccessory, itemRegistry, updater,
-                        BatteryStatus.getFromCharacteristics(characteristicItems));
-            case LOCK:
-                return new HomekitLockImpl(taggedItem, itemRegistry, updater);
+        try {
+            final Class<? extends AbstractHomekitAccessoryImpl> accessoryImplClass = serviceImplMap
+                    .get(taggedItem.getAccessoryType());
+            if (accessoryImplClass != null) {
+                accessoryImpl = accessoryImplClass
+                        .getConstructor(HomekitTaggedItem.class, List.class, HomekitAccessoryUpdater.class,
+                                HomekitSettings.class)
+                        .newInstance(taggedItem, requiredCharacteristics, updater, settings);
+                addOptionalCharacteristics(accessoryImpl, metadataRegistry);
+                return accessoryImpl;
+            } else {
+                LOGGER.error("Unsupported HomeKit type: {}", taggedItem.getAccessoryType());
+                throw new HomekitException("Unsupported HomeKit type: " + taggedItem.getAccessoryType());
+            }
+        } catch (NoSuchMethodException | IllegalAccessException | InstantiationException
+                | InvocationTargetException e) {
+            LOGGER.error("Cannot instantiate accessory implementation for accessory {}", taggedItem.getAccessoryType(),
+                    e);
+            throw new HomekitException(
+                    "Cannot instantiate accessory implementation for accessory " + taggedItem.getAccessoryType());
         }
+    }
+
+    /**
+     * return HomeKit accessory types for a OH item based on meta data
+     * 
+     * @param item OH item
+     * @param metadataRegistry meta data registry
+     * @return list of HomeKit accessory types and characteristics.
+     */
+    public static List<Entry<HomekitAccessoryType, HomekitCharacteristicType>> getAccessoryTypes(Item item,
+            MetadataRegistry metadataRegistry) {
+        final List<Entry<HomekitAccessoryType, HomekitCharacteristicType>> accessories = new ArrayList<>();
+        Metadata metadata = metadataRegistry.get(new MetadataKey(METADATA_KEY, item.getUID()));
+        boolean legacyMode = !(metadata != null);
+        String[] tags = !legacyMode ? metadata.getValue().split(",") : item.getTags().toArray(new String[0]); // fallback
+                                                                                                              // to tags
+
+        LOGGER.debug("item {} meta data {}  tags {} ", item.getName(), metadata, tags);
+        for (String tag : tags) {
```

Comment:
While iterating over the tags you could check if we are in `legacyMode` and adjust according to the old behaviour if the tag "Lighting" is found.

Reply:
ah, ok.. missed that one with overwriting. we could add the same smart logic but personally i would prefer to have full control over the characteristics exposed via homekit. 
but you are right, for legacy item we should support this smart logic.

Reply:
Yes, only for legacy tags.

Reply:
done